lark/parser_frontends.py

```diff
@@ -93,26 +99,88 @@ def _verify_start(self, start=None):
+        #   passed `None` with regard to start_pos and end_pos
         chosen_start = self._verify_start(start)
         if self.parser_conf.parser_type != 'lalr':
             raise ConfigurationError("parse_interactive() currently only works with parser='lalr' ")
-        stream = self._make_lexer_thread(text)  # type: ignore[arg-type]
+        stream = self._make_lexer_thread(text, start_pos=start_pos, end_pos=end_pos)  # type: ignore[arg-type]
         return self.parser.parse_interactive(stream, chosen_start)
 
 
+    def scan(self, text: str, start: Optional[str]=None, *, start_pos: Optional[int] = None,
+             end_pos: Optional[int] = None):
+        """
+        In contrast to the other functions here, this one actually does work. See `Lark.scan`
+        for a description of what this function is for.
+        """
+        if self.options.parser != 'lalr':
+            raise ValueError("scan requires parser='lalr' and lexer='contextual'")
+        start_states = self.parser._parse_table.start_states
+        chosen_start = self._verify_start(start)
+        start_state = start_states[chosen_start]
+        pos = start_pos if start_pos is not None else 0
+        end_pos = end_pos if end_pos is not None else len(text)
+        if pos < 0:
+            pos += len(text)
+        if end_pos < 0:
+            pos += len(text)
+        del start_pos
+        while True:
+            # Find the next candidate location
+            found = self.lexer.search_start(text, start_state, pos, end_pos)
+            # No more valid candidates
+            if found is None:
+                break
+            assert found.end_pos <= end_pos
+            # Collect the potential end points found for this parse
+            # We need to keep track of multiple options in case there are false `$END`s in the `ip.choices()`
+            # We don't want to check early since this can be expensive.
+            valid_end = []
+            ip = self.parse_interactive(text, start=chosen_start, start_pos=found.start_pos, end_pos=end_pos)
+            tokens = ip.lexer_thread.lex(ip.parser_state)
+            while True:
+                try:
+                    token = next(tokens)
+                    ip.feed_token(token)
+                except (UnexpectedInput, StopIteration):
+                    # Either we couldn't parse the characters or the resulting token wasn't valid.
+                    # Either way, stop
+                    break
+                if '$END' in ip.choices():
+                    valid_end.append((token, ip.copy()))
+            # Check through all potential ending points and see if passing in `$END` actually works
+            for (last, pot) in valid_end[::-1]:
+                try:
+                    res = pot.feed_eof(last)
+                except UnexpectedInput:
+                    continue
+                else:
+                    yield ((found.start_pos, last.end_pos), res)
```

Comment:
We'll probably want scan to yield a ScanMatch (like re does), and not just a tuple.

Reply:
I don't know what you mean with `ScanMatch`? Should I just introduce a named tuple? (I do like the ability to unpack it easily in a loop, so I would be against to something other than a namedtuple)

Reply:
NamedTuple is okay too

Reply:
`for (start_pos, end_pos), tree in parse.scan(text)` is the syntax I would like to allow. That requires `ScanMatch` to be iterable, and a namedtuple is the easiest way to achieve this without manually implementing an `__iter__` method.

Reply:
Moved to a `NamedTuple` defined in `parser_frontends.py`